File: App/lib/services/symptom_triage.dart
import '../models/doctor.dart';

/// Deterministic symptom -> specialty routing.
///
/// This is a lightweight triage HEURISTIC, not a diagnosis. The UI must show a
/// clear "not medical advice" disclaimer. It maps free-text symptoms to one or
/// more specialties and always falls back to a General Physician.
class SymptomTriage {
  const SymptomTriage();

  /// Canonical specialty registry (single source of truth).
  static const List<Specialty> specialties = [
    Specialty(
      key: 'general',
      label: 'General Physician',
      searchKeyword: 'general physician clinic doctor',
      nameHints: ['general', 'family', 'physician', 'medicine'],
    ),
    Specialty(
      key: 'cardiology',
      label: 'Cardiologist',
      searchKeyword: 'cardiologist heart specialist clinic',
      nameHints: ['cardio', 'heart'],
    ),
    Specialty(
      key: 'neurology',
      label: 'Neurologist',
      searchKeyword: 'neurologist brain nerve specialist clinic',
      nameHints: ['neuro', 'brain', 'nerve'],
    ),
    Specialty(
      key: 'dermatology',
      label: 'Dermatologist',
      searchKeyword: 'dermatologist skin specialist clinic',
      nameHints: ['derma', 'skin'],
    ),
    Specialty(
      key: 'gastroenterology',
      label: 'Gastroenterologist',
      searchKeyword: 'gastroenterologist stomach digestive specialist clinic',
      nameHints: ['gastro', 'digest'],
    ),
    Specialty(
      key: 'orthopedics',
      label: 'Orthopedic Surgeon',
      searchKeyword: 'orthopedic bone joint specialist clinic',
      nameHints: ['ortho', 'bone', 'joint'],
    ),
    Specialty(
      key: 'ent',
      label: 'ENT Specialist',
      searchKeyword: 'ENT ear nose throat specialist clinic',
      nameHints: ['ent', 'ear', 'nose', 'throat'],
    ),
    Specialty(
      key: 'ophthalmology',
      label: 'Ophthalmologist',
      searchKeyword: 'ophthalmologist eye specialist clinic',
      nameHints: ['ophthal', 'eye', 'vision'],
    ),
    Specialty(
      key: 'pediatrics',
      label: 'Pediatrician',
      searchKeyword: 'pediatrician child specialist clinic',
      nameHints: ['pediat', 'child', 'kids'],
    ),
    Specialty(
      key: 'gynecology',
      label: 'Gynecologist',
      searchKeyword: 'gynecologist obstetrician women clinic',
      nameHints: ['gyne', 'obstet', 'women', 'maternity'],
    ),
    Specialty(
      key: 'urology',
      label: 'Urologist',
      searchKeyword: 'urologist kidney urinary specialist clinic',
      nameHints: ['uro', 'kidney'],
    ),
    Specialty(
      key: 'psychiatry',
      label: 'Psychiatrist',
      searchKeyword: 'psychiatrist mental health specialist clinic',
      nameHints: ['psychi', 'mental'],
    ),
    Specialty(
      key: 'endocrinology',
      label: 'Endocrinologist',
      searchKeyword: 'endocrinologist diabetes thyroid hormone specialist clinic',
      nameHints: ['endocrin', 'diabet', 'thyroid', 'hormone'],
    ),
    Specialty(
      key: 'pulmonology',
      label: 'Pulmonologist',
      searchKeyword: 'pulmonologist lung chest specialist clinic',
      nameHints: ['pulmo', 'lung', 'chest', 'respir'],
    ),
    Specialty(
      key: 'dentistry',
      label: 'Dentist',
      searchKeyword: 'dentist dental clinic',
      nameHints: ['dental', 'dentist', 'tooth'],
    ),
  ];

  static Specialty get general =>
      specialties.firstWhere((s) => s.key == 'general');

  static Specialty? byKey(String key) {
    for (final s in specialties) {
      if (s.key == key) return s;
    }
    return null;
  }

  /// Symptom keyword -> specialty key. Kept simple and readable on purpose;
  /// extend freely.
  static const Map<String, String> _symptomToSpecialty = {
    'chest pain': 'cardiology',
    'palpitation': 'cardiology',
    'heart': 'cardiology',
    'breathless': 'pulmonology',
    'shortness of breath': 'pulmonology',
    'cough': 'pulmonology',
    'wheeze': 'pulmonology',
    'headache': 'neurology',
    'migraine': 'neurology',
    'dizziness': 'neurology',
    'seizure': 'neurology',
    'numbness': 'neurology',
    'rash': 'dermatology',
    'acne': 'dermatology',
    'itch': 'dermatology',
    'skin': 'dermatology',
    'stomach': 'gastroenterology',
    'abdominal': 'gastroenterology',
    'nausea': 'gastroenterology',
    'diarrhea': 'gastroenterology',
    'vomit': 'gastroenterology',
    'acidity': 'gastroenterology',
    'bone': 'orthopedics',
    'joint': 'orthopedics',
    'fracture': 'orthopedics',
    'back pain': 'orthopedics',
    'knee': 'orthopedics',
    'ear': 'ent',
    'nose': 'ent',
    'throat': 'ent',
    'sinus': 'ent',
    'eye': 'ophthalmology',
    'vision': 'ophthalmology',
    'blurry': 'ophthalmology',
    'child': 'pediatrics',
    'infant': 'pediatrics',
    'baby': 'pediatrics',
    'pregnan': 'gynecology',
    'menstru': 'gynecology',
    'period': 'gynecology',
    'urine': 'urology',
    'urinary': 'urology',
    'kidney': 'urology',
    'anxiety': 'psychiatry',
    'depression': 'psychiatry',
    'stress': 'psychiatry',
    'mental': 'psychiatry',
    'diabetes': 'endocrinology',
    'sugar': 'endocrinology',
    'thyroid': 'endocrinology',
    'tooth': 'dentistry',
    'gum': 'dentistry',
    'dental': 'dentistry',
    'fever': 'general',
    'cold': 'general',
    'flu': 'general',
    'body ache': 'general',
    'weakness': 'general',
  };

  /// Returns matching specialties ordered by how many symptom cues hit them.
  /// Always non-empty (falls back to General Physician).
  List<Specialty> analyzeSymptoms(String text) {
    final t = text.toLowerCase();
    final tally = <String, int>{};
    _symptomToSpecialty.forEach((symptom, key) {
      if (t.contains(symptom)) {
        tally[key] = (tally[key] ?? 0) + 1;
      }
    });

    if (tally.isEmpty) return [general];

    final keys = tally.keys.toList()
      ..sort((a, b) => tally[b]!.compareTo(tally[a]!));
    return keys.map((k) => byKey(k)!).toList();
  }
}

File: App/app/src/main/java/com/example/ui/SymptomTriage.kt
package com.example.ui

/**
 * Lightweight symptom -> specialty routing, plus the Places search keyword and
 * name hints for each specialty.
 *
 * This is a triage HEURISTIC, not a diagnosis — the UI must show a
 * "guidance only" disclaimer. Unrecognised input always falls back to GENERAL.
 *
 * Note: Google Places cannot return individual named doctors, their specialties
 * or schedules. Doctor mode therefore locates CLINICS / SPECIALIST PRACTICES /
 * HOSPITALS matching a specialty — it never claims to list named physicians.
 */
object SymptomTriage {

    /** Places query used to find places for this specialty. */
    fun searchKeyword(specialty: DoctorSpecialty): String = when (specialty) {
        DoctorSpecialty.GENERAL -> "general physician clinic doctor"
        DoctorSpecialty.CARDIOLOGIST -> "cardiologist heart specialist clinic"
        DoctorSpecialty.DERMATOLOGIST -> "dermatologist skin specialist clinic"
        DoctorSpecialty.ORTHOPEDIC -> "orthopedic bone joint specialist clinic"
        DoctorSpecialty.ENT -> "ENT ear nose throat specialist clinic"
        DoctorSpecialty.GYNECOLOGIST -> "gynecologist obstetrician women clinic"
        DoctorSpecialty.PEDIATRICIAN -> "pediatrician child specialist clinic"
        DoctorSpecialty.NEUROLOGIST -> "neurologist brain nerve specialist clinic"
        DoctorSpecialty.PSYCHIATRIST -> "psychiatrist mental health clinic"
        DoctorSpecialty.OPHTHALMOLOGIST -> "ophthalmologist eye specialist clinic"
        DoctorSpecialty.DENTIST -> "dentist dental clinic"
        DoctorSpecialty.UROLOGIST -> "urologist kidney urinary specialist clinic"
    }

    /** Words in a place name that signal this specialty (used to boost ranking). */
    fun nameHints(specialty: DoctorSpecialty): List<String> = when (specialty) {
        DoctorSpecialty.GENERAL -> listOf("general", "family", "physician", "medicine")
        DoctorSpecialty.CARDIOLOGIST -> listOf("cardio", "heart")
        DoctorSpecialty.DERMATOLOGIST -> listOf("derma", "skin")
        DoctorSpecialty.ORTHOPEDIC -> listOf("ortho", "bone", "joint")
        DoctorSpecialty.ENT -> listOf("ent", "ear", "nose", "throat")
        DoctorSpecialty.GYNECOLOGIST -> listOf("gyne", "obstet", "women", "maternity")
        DoctorSpecialty.PEDIATRICIAN -> listOf("pediat", "child", "kids", "shishu")
        DoctorSpecialty.NEUROLOGIST -> listOf("neuro", "brain", "nerve")
        DoctorSpecialty.PSYCHIATRIST -> listOf("psychi", "mental")
        DoctorSpecialty.OPHTHALMOLOGIST -> listOf("ophthal", "eye", "vision")
        DoctorSpecialty.DENTIST -> listOf("dental", "dentist", "tooth")
        DoctorSpecialty.UROLOGIST -> listOf("uro", "kidney")
    }

    /** Symptom keyword -> specialty. Extend freely. */
    private val symptomMap: Map<String, DoctorSpecialty> = mapOf(
        "chest pain" to DoctorSpecialty.CARDIOLOGIST,
        "palpitation" to DoctorSpecialty.CARDIOLOGIST,
        "heart" to DoctorSpecialty.CARDIOLOGIST,
        "blood pressure" to DoctorSpecialty.CARDIOLOGIST,

        "headache" to DoctorSpecialty.NEUROLOGIST,
        "migraine" to DoctorSpecialty.NEUROLOGIST,
        "dizziness" to DoctorSpecialty.NEUROLOGIST,
        "seizure" to DoctorSpecialty.NEUROLOGIST,
        "numbness" to DoctorSpecialty.NEUROLOGIST,
        "stroke" to DoctorSpecialty.NEUROLOGIST,

        "rash" to DoctorSpecialty.DERMATOLOGIST,
        "acne" to DoctorSpecialty.DERMATOLOGIST,
        "itch" to DoctorSpecialty.DERMATOLOGIST,
        "skin" to DoctorSpecialty.DERMATOLOGIST,
        "eczema" to DoctorSpecialty.DERMATOLOGIST,

        "bone" to DoctorSpecialty.ORTHOPEDIC,
        "joint" to DoctorSpecialty.ORTHOPEDIC,
        "fracture" to DoctorSpecialty.ORTHOPEDIC,
        "back pain" to DoctorSpecialty.ORTHOPEDIC,
        "knee" to DoctorSpecialty.ORTHOPEDIC,
        "shoulder" to DoctorSpecialty.ORTHOPEDIC,

        "ear" to DoctorSpecialty.ENT,
        "nose" to DoctorSpecialty.ENT,
        "throat" to DoctorSpecialty.ENT,
        "sinus" to DoctorSpecialty.ENT,
        "tonsil" to DoctorSpecialty.ENT,

        "eye" to DoctorSpecialty.OPHTHALMOLOGIST,
        "vision" to DoctorSpecialty.OPHTHALMOLOGIST,
        "blurry" to DoctorSpecialty.OPHTHALMOLOGIST,

        "pregnan" to DoctorSpecialty.GYNECOLOGIST,
        "menstru" to DoctorSpecialty.GYNECOLOGIST,
        "period pain" to DoctorSpecialty.GYNECOLOGIST,

        "child" to DoctorSpecialty.PEDIATRICIAN,
        "infant" to DoctorSpecialty.PEDIATRICIAN,
        "baby" to DoctorSpecialty.PEDIATRICIAN,

        "urine" to DoctorSpecialty.UROLOGIST,
        "urinary" to DoctorSpecialty.UROLOGIST,
        "kidney" to DoctorSpecialty.UROLOGIST,

        "anxiety" to DoctorSpecialty.PSYCHIATRIST,
        "depress" to DoctorSpecialty.PSYCHIATRIST,
        "stress" to DoctorSpecialty.PSYCHIATRIST,
        "insomnia" to DoctorSpecialty.PSYCHIATRIST,

        "tooth" to DoctorSpecialty.DENTIST,
        "teeth" to DoctorSpecialty.DENTIST,
        "gum" to DoctorSpecialty.DENTIST,
        "dental" to DoctorSpecialty.DENTIST,

        "fever" to DoctorSpecialty.GENERAL,
        "cold" to DoctorSpecialty.GENERAL,
        "flu" to DoctorSpecialty.GENERAL,
        "cough" to DoctorSpecialty.GENERAL,
        "body ache" to DoctorSpecialty.GENERAL,
        "weakness" to DoctorSpecialty.GENERAL,
        "vomit" to DoctorSpecialty.GENERAL,
        "diarrhea" to DoctorSpecialty.GENERAL,
        "stomach" to DoctorSpecialty.GENERAL
    )

    /**
     * Returns specialties matching the free-text symptoms, most-cued first.
     * Empty list when nothing is recognised (caller decides the fallback).
     */
    fun analyze(text: String): List<DoctorSpecialty> {
        if (text.isBlank()) return emptyList()
        val t = text.lowercase()
        val tally = mutableMapOf<DoctorSpecialty, Int>()
        symptomMap.forEach { (symptom, specialty) ->
            if (t.contains(symptom)) tally[specialty] = (tally[specialty] ?: 0) + 1
        }
        return tally.entries.sortedByDescending { it.value }.map { it.key }
    }
}
